service/worker/scanner/executions/shard_validator.go

```diff
@@ -0,0 +1,358 @@
+		mutableState.ExecutionInfo,
+		mutableState.ExecutionState,
+		shardID,
+		limiter,
+		historyDB,
+		dbRequestsCount,
+		execMgr,
+	)
+	if !executionValidationResult.isValid {
+		return newExecutionValidationFailureWithFailures(
+			executionData, map[string]executionValidatorResult{"history_fetcher": executionValidationResult},
+		)
+	}
+	if executionData.history == nil || executionData.historyBranch == nil {
+		return nil
+	}
+
+	var result *executionValidationFailure = nil
+	for _, executionValidator := range validators {
+		executionValidationResult = executionValidator.validate(executionData)
+		if !executionValidationResult.isValid {
+			if result == nil {
+				result = newExecutionValidationFailure(executionData)
+			}
+			if result.Failures == nil {
+				result.Failures = map[string]executionValidatorResult{}
+			}
+			result.Failures[executionValidator.validatorTag()] = executionValidationResult
+		}
+	}
+	return result
+}
+
+// Can return NoFailure, nil, nil if execution was removed together with history branch.
+func fetchHistoryBranch(
+	executionInfo *persistencespb.WorkflowExecutionInfo,
+	executionState *persistencespb.WorkflowExecutionState,
+	shardID int32,
+	limiter quotas.RateLimiter,
+	historyDB persistence.HistoryManager,
+	totalDBRequests *int64,
+	execMgr persistence.ExecutionManager,
+) (executionValidatorResult, []*historypb.HistoryEvent, *persistencespb.HistoryBranch) {
+	var branch *persistencespb.HistoryBranch
+	currentVersionHistory, err := versionhistory.GetCurrentVersionHistory(executionInfo.VersionHistories)
+	if err != nil {
+		return executionValidatorResult{
+			isValid:               false,
+			failureReasonTag:      "failed_fetch_history",
+			failureReasonDetailed: "Failed to fetch current version history",
+		}, nil, nil
+	}
+
+	branch, err = serialization.HistoryBranchFromBlob(currentVersionHistory.BranchToken,
+		enumspb.ENCODING_TYPE_PROTO3.String())
+
+	if err != nil {
+		return executionValidatorResult{
+			isValid:               false,
+			failureReasonTag:      "failed_decode_branch_token",
+			failureReasonDetailed: "Failed to decode current branch token.",
+		}, nil, nil
+	}
+
+	var historyPageSize int = 1
+	readHistoryBranchReq := &persistence.ReadHistoryBranchRequest{
+		MinEventID:    common.FirstEventID,
+		MaxEventID:    common.EndEventID,
+		BranchToken:   currentVersionHistory.BranchToken,
+		ShardID:       &shardID,
+		PageSize:      historyPageSize,
+		NextPageToken: nil,
+	}
+
+	preconditionForDBCall(totalDBRequests, limiter)
+	history, err := historyDB.ReadHistoryBranch(readHistoryBranchReq)
+
+	if err != nil {
+		if errors.Is(err, gocql.ErrNotFound) {
+			stillExists, failureInfo := concreteExecutionExists(executionInfo, executionState, execMgr, limiter, totalDBRequests)
+			if !failureInfo.isValid {
+				return failureInfo, nil, nil
+			}
+			if !stillExists {
+				return executionValidatorResultNoCorruption, nil, nil
+			}
```

Comment:
seems that https://github.com/temporalio/temporal/pull/1244/files#diff-67d79a77b4f54bf32b00b7a2f66d250529578c850b45de4c76e2be4adaeda9a7R141 already loaded the mutable state from DB, so why checking again?